File: latexdraw-core/net.sf.latexdraw/src/main/net/sf/latexdraw/actions/shape/CopyShapes.scala
package net.sf.latexdraw.actions.shape

import java.util.ArrayList

import scala.collection.JavaConversions.asScalaBuffer

import org.malai.action.Action

import net.sf.latexdraw.glib.models.interfaces.DrawingTK
import net.sf.latexdraw.glib.models.interfaces.IShape

/**
 * This action copies the selected shapes.<br>
 * <br>
 * This file is part of LaTeXDraw.<br>
 * Copyright (c) 2005-2013 Arnaud BLOUIN<br>
 * <br>
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * <br>
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.<br>
 * <br>
 * 2012-04-20<br>
 * @author Arnaud BLOUIN
 * @since 3.0
 */
class CopyShapes extends Action {
	/** The selection action. */
	protected[shape] var _selection : Option[SelectShapes] = None

	/** The copied shapes from the selection. */
	private[shape] var copiedShapes : java.util.List[IShape] = new ArrayList[IShape]()

	/** The number of times that the shapes have been copied. Use to compute the gap while pasting. */
	private[shape] var nbTimeCopied : Int = 0


	override protected def doActionBody() {
		_selection.get.shapes.foreach{shape => copiedShapes.add(DrawingTK.getFactory.duplicate(shape))}
	}


	override def cancelledBy(action : Action) = action.isInstanceOf[CopyShapes]


	override def isRegisterable() = true


	override def canDo() = _selection.isDefined && !_selection.get.shapes.isEmpty


	/**
	 * @param selection The selected shapes to copy or cut.
	 */
	def setSelection(selection : SelectShapes) {
		_selection = Some(selection)
	}

	def selection = _selection


	override def flush() {
		super.flush
		copiedShapes.clear
		_selection 	 = None
	}
}

File: latexdraw-core/net.sf.latexdraw/src/main/net/sf/latexdraw/actions/shape/CopyShapes.java
/*
 * This file is part of LaTeXDraw.
 * Copyright (c) 2005-2016 Arnaud BLOUIN
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */
package net.sf.latexdraw.actions.shape;

import java.util.List;
import java.util.stream.Collectors;
import net.sf.latexdraw.models.ShapeFactory;
import net.sf.latexdraw.models.interfaces.shape.IShape;
import org.malai.action.Action;
import org.malai.action.ActionImpl;

/**
 * This action copies the selected shapes.
 * @author Arnaud BLOUIN
 */
public class CopyShapes extends ActionImpl {
	/** The selection action. */
	protected SelectShapes selection;

	/** The copied shapes from the selection. */
	protected List<IShape> copiedShapes;

	/** The number of times that the shapes have been copied. Use to compute the gap while pasting. */
	protected int nbTimeCopied;


	@Override
	protected void doActionBody() {
		copiedShapes = selection.getShapes().stream().map(ShapeFactory.INST::duplicate).collect(Collectors.toList());
	}

	@Override
	public boolean cancelledBy(final Action action) {
		return action instanceof CopyShapes;
	}

	@Override
	public boolean isRegisterable() {
		return true;
	}

	@Override
	public boolean canDo() {
		return selection != null && !selection.getShapes().isEmpty();
	}

	/**
	 * @param sel The selected shapes to copy or cut.
	 */
	public void setSelection(final SelectShapes sel) {
		selection = sel;
	}

	public SelectShapes getSelection() {
		return selection;
	}


	@Override
	public void flush() {
		super.flush();
		copiedShapes.clear();
		selection = null;
	}
}
